### PR TITLE
feature/mx-1632 structured file support

### PR DESCRIPTION
**Pull request by ZehraVictoria**
# PR Context

So far mex drop only allows JSON file upload.
This establishes compatibility with most common structured file formats.

---
# Added

- support for more structured file formats

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -3,3 +3,6 @@ mex-release @ git+https://github.com/robert-koch-institut/mex-release.git
 pdm==2.17.3
 pre-commit==3.8.0
 wheel==0.44.0
+openpyxl==3.0.10
+
+reflex==0.5.8
```
Comment (mr-kamran-ali) [inline requirements.txt:8]: We have reflex in `pyproject.toml` this is being added here by some hook, which is also modifying `gitignore`. its better to remove `reflex` from here.

File: .gitignore
Comment (mr-kamran-ali) [inline .gitignore:1]: our `gitignore` structure was better. this is being modified by some hook. Can you please restore it?

File: mex/drop/api.py [hunk 2]
```diff
@@ -146,18 +187,49 @@ async def drop_data_multipart(
...
+        elif content_type == "application/xml":
+            decoded_data = content.decode("utf-8")
+            with open(out_file, "w") as f:
+                f.write(decoded_data)
+        elif (
+            content_type == "application/vnd.ms-excel"
+            or content_type
+            == "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
+        ):
+            file_stream = io.BytesIO(content)
```
Comment (mr-kamran-ali) [inline mex/drop/api.py:209]: this `elif` block is not being tested. thats why total coverage is not reaching 95%. can you verify its being tested in calling endpoints?

**Review by mr-kamran-ali — COMMENTED** [review]
Nice work. Few little things to fix.

File: tests/test_main.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 
 @pytest.mark.integration()
+@pytest.mark.skip(reason="features not implemented yet")
```
Comment (mr-kamran-ali) [inline tests/test_main.py:6]: this is what causing the screen shot upload action to fail. this test is skipped so no screen shots are created. Can you remove this line and see if all actions pass?

Comment (ZehraVictoria) [reply tests/test_main.py]: Thanks, that absolutely fixed it!

**Review by rababerladuseladim — CHANGES_REQUESTED** [review]
Nice work! I do have a few remarks.

First and foremost: One test fails locally: `tests/test_main.py::test_index[firefox]`

File: tests/test_api.py [hunk 4]
```diff
@@ -116,8 +305,8 @@ def test_drop_multiple_files(
     settings: DropSettings,
 ) -> None:
     files_data = [
-        ("files", (name, BytesIO(content.encode()), "text/plain"))
-        for name, content in files.items()
+        ("files", (name, BytesIO(content.encode()), content_type))
+        for name, (content, content_type) in files.items()
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply tests/test_api.py:309]: That would make `files_data` incompatible with the` files` parameter from `client.post` (Exact error: Argument of type "list[tuple[str, BytesIO, str]]" cannot be assigned to parameter "files" of type "RequestFiles | None" in function "post")

File: pyproject.toml [hunk 5]
```diff
@@ -10,6 +10,10 @@ requires-python = "<3.12,>=3.11"
 dependencies = [
     "mex-common @ git+https://github.com/robert-koch-institut/mex-common.git@0.35.0",
     "reflex>=0.5.9,<1",
+    "mex-common @ git+https://github.com/robert-koch-institut/mex-common.git@0.35.0",
+    "openpyxl>=3.0.10,<4",
+    "reflex>=0.5.9,<1",
```
Comment (rababerladuseladim) [inline pyproject.toml:15]: mex-common and reflex are still duplicates, please remove. And what is playright used for? Isn't it only a testing dependency? If yes: please move to dev-dependencies.

File: mex/drop/api.py [hunk 6]
```diff
@@ -158,6 +204,21 @@ async def write_to_file(content: bytes, out_file: pathlib.Path) -> None:
         ) from exc
 
 
+async def validate_file_extension(content_type: str | None, filename: str) -> None:
+    """Validate uploaded file format."""
+    if content_type not in ALLOWED_CONTENT_TYPES:
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail=f"Unsupported file extension: {type}",
```
Comment (rababerladuseladim) [inline mex/drop/api.py:212]: This is also a bad request, isn't it? And the error could contain the allowed content types, to make the life of the users easier.

But after a doing some research, I found an even better error code: [422 Unprocessable Content](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422)

**Comment by rababerladuseladim** [comment]
And one more: could you document the allowed file types in swagger as well? That makes using the API easier.

File: mex/drop/api.py [hunk 7]
```diff
@@ -170,58 +167,56 @@ async def drop_data_multipart(
...
+def check_duplicate_filenames(files: list[UploadFile]) -> None:
+    """Check for duplicate filenames."""
+    hash_bucket = set()
+    for file in files:
+        if file.filename in hash_bucket:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="Duplicate filename.",
+            )
+        hash_bucket.add(file.filename)
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply mex/drop/api.py]: Used in 5c69fc5

File: mex/drop/api.py [hunk 8]
```diff
@@ -110,19 +114,13 @@ async def drop_data(
...
-        )
-    raw_data = await request.body()
-    return Response(
-        status_code=200,
-        background=BackgroundTask(write_to_file, raw_data, out_file, content_type),
-    )
+    out_file = settings.drop_directory / x_system / f"{entity_type}{file_ext}"
+    if content_type == "application/json" and isinstance(data, (dict | list)):
+        background_tasks.add_task(json_sink, data, out_file)
+        return Response(status_code=200)
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply mex/drop/api.py]: Fixed in 5c69fc5

File: mex/drop/api.py [hunk 9]
```diff
@@ -62,33 +60,40 @@ async def drop_data(
...
+    background_tasks: BackgroundTasks,
 ) -> Response:
-    """Upload arbitrary structured data to MEx.
+    """Upload arbitrary data to MEx.
 
     Args:
         x_system: name of the x-system that the data comes from
         entity_type: name of the data file that is uploaded, if unsure use 'default'
-        data: dictionary with string keys or list with and arbitrary values
+        data: data content of request body (dict, list or bytes)
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply mex/drop/api.py]: Fixed in 5c69fc5

File: mex/drop/api.py [hunk 10]
```diff
@@ -62,33 +60,40 @@ async def drop_data(
...
         Body(
             description=(
                 "An arbitrary data structure, that can be further processed by MEx"
             ),
             examples=[
                 {"foo": "bar", "list": [1, 2, "foo"], "nested": {"foo": "bar"}},
+                [{"foo": "bar"}, {"bar": [1, 2, 3]}],
                 "foo,bar\n1,2",
                 '{"foo": "bar"}',
                 "<root><foo>bar</foo></root>",
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply mex/drop/api.py]: Fixed in 5c69fc5

File: pyproject.toml (hunk 5 above)
Comment (ZehraVictoria) [reply pyproject.toml]: rm duplicates and playright in d74f846b5af0128f26637f63dea8c13ab09ab922

File: tests/test_api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply tests/test_api.py:304]: fixed in 5c69fc5e73f0eeb983ff7935cbb09d0e9b6bf1e8

File: tests/test_api.py [hunk 11]
```diff
@@ -60,27 +182,57 @@ def test_drop_data(
...
+            headers=(
+                {"X-API-Key": api_key, "Content-Type": content_type} if api_key else {}
+            ),
+            content=expected_content,
+        )
     assert response.status_code == expected_response_code, response.text
+    if content_type in ALLOWED_CONTENT_TYPES:
+        expected_file = Path(
+            settings.drop_directory, x_system, entity_type
+        ).with_suffix(ALLOWED_CONTENT_TYPES[content_type])
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply tests/test_api.py]: fixed in 5c69fc5e73f0eeb983ff7935cbb09d0e9b6bf1e8

File: mex/drop/api.py (hunk 6 above)
Comment (ZehraVictoria) [reply mex/drop/api.py]: fixed in 96353b6bae23b9886e76323a53f1b33d07f0aa41

File: pyproject.toml [hunk 12]
```diff
@@ -10,10 +10,15 @@ requires-python = "<3.12,>=3.11"
...
+    "playwright>=1.46.0",
 ]
 optional-dependencies.dev = [
     "ipdb>=0.13.13,<1",
     "mypy>=1.11.1,<2",
+    "pandas-stubs>=2.2.2.240807,<3",
```
[the remark this answers is not in the document]
Comment (ZehraVictoria) [reply pyproject.toml]: I actually removed the pandas-stubs, instead of openpyxl since that is still needed for testing in the current code.

[the remark this answers is not in the document]
Comment (rababerladuseladim) [reply pyproject.toml]: well, true, But since we do not parse it, a bytes comparison without pandas would be sufficient for me. Let's fix this when bored, though.

File: mex/drop/api.py [hunk 13]
```diff
@@ -110,19 +110,12 @@ async def drop_data(
...
-    return Response(
-        status_code=200,
-        background=BackgroundTask(write_to_file, raw_data, out_file, content_type),
-    )
+    out_file = settings.drop_directory / x_system / f"{entity_type}{file_ext}"
+    if content_type == "application/json" and isinstance(data, (dict | list)):
+        background_tasks.add_task(json_sink, data, out_file)
+    if isinstance(data, bytes):
+        background_tasks.add_task(write_to_file, data, out_file)
+    return Response(status_code=200)
```
Comment (rababerladuseladim) [inline mex/drop/api.py:118]: Just a question: why do you not add the background task to the Response anymore?

File: mex/drop/api.py [hunk 14]
```diff
@@ -170,60 +167,16 @@ async def drop_data_multipart(
             detail="API Key not authorized to drop data for this x_system.",
         )
     settings = DropSettings.get()
+    check_duplicate_filenames(files)
```
Comment (rababerladuseladim) [inline mex/drop/api.py:170]: Could you add a test case with duplicate filenames?

File: mex/drop/files_io.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            detail=(
+                f"Unsupported content type: {content_type}. "
+                f"Allowed types: {', '.join(ALLOWED_CONTENT_TYPES.values())}"
+            ),
+        )
+
+    suffix = pathlib.Path(filename).suffix
+    if ALLOWED_CONTENT_TYPES[content_type] != suffix and suffix != ".csv":
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
```
Comment (rababerladuseladim) [inline mex/drop/files_io.py:58]: This should also be a 422, shouldn't it? Same for line 67.

**Review by rababerladuseladim — CHANGES_REQUESTED** [review]
Biu-ti-ful! And almost there, just some tiny nitpicking left.

File: mex/drop/api.py (hunk 13 above)
Comment (ZehraVictoria) [reply mex/drop/api.py:118]: no reason really. Happy to change it back :)

File: mex/drop/api.py [hunk 16]
```diff
@@ -14,7 +14,7 @@
 )
 from pydantic import BaseModel
 from starlette import status
-from starlette.background import BackgroundTasks
+from starlette.background import BackgroundTask, BackgroundTasks
```
Comment (rababerladuseladim) [inline mex/drop/api.py:17]: Now you do not need the `BackgroundTasks` anmore, do you?

**Review by rababerladuseladim — CHANGES_REQUESTED** [review]
oh boi, just one tiny thing left.

File: mex/drop/api.py (hunk 16 above)
Comment (ZehraVictoria) [reply mex/drop/api.py:17]: `BackgroundTasks` is still used in multipart upload. but I removed it from the single file drop function: 86eec1301f7640b7e5209c1d755733ef59df1431
